File: src/ProtonVPN.InstallActions/SystemState.cpp
#include "pch.h"
#include "winreg.h"

HKEY get_key(const LPCWSTR key)
{
	HKEY h_key = nullptr;
	RegOpenKeyEx(HKEY_LOCAL_MACHINE, key, 0, KEY_READ | KEY_WOW64_64KEY, &h_key);

	return h_key;
}

bool run_once_check()
{
	const auto key = get_key(TEXT("SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\RunOnce"));
	if (key != nullptr)
	{
		WCHAR sz_buffer[512];
		DWORD dw_buffer_size = sizeof sz_buffer;
		const auto result = RegQueryValueEx(key, nullptr, nullptr, nullptr, reinterpret_cast<LPBYTE>(sz_buffer),
		                                    &dw_buffer_size);
		if (result == ERROR_SUCCESS && wcscmp(sz_buffer, L"DVDRebootSignal") == 0)
		{
			RegCloseKey(key);
			return true;
		}

		RegCloseKey(key);
	}

	return false;
}

bool check_keys_exist()
{
	LPCWSTR keys_exist[] = {
		TEXT("SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Component Based Servicing\\RebootPending"),
		TEXT("SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Component Based Servicing\\RebootInProgress"),
		TEXT("SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\WindowsUpdate\\Auto Update\\RebootRequired"),
		TEXT("SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Component Based Servicing\\PackagesPending"),
		TEXT("SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\WindowsUpdate\\Auto Update\\PostRebootReporting"),
	};

	for (const auto key : keys_exist)
	{
		const auto key_p = get_key(key);
		if (key_p != nullptr)
		{
			RegCloseKey(key_p);
			return true;
		}
	}

	return false;
}

bool pending_update()
{
	const auto key = get_key(TEXT("SOFTWARE\\Microsoft\\Updates"));
	if (key != nullptr)
	{
		WCHAR sz_buffer[512];
		DWORD dw_buffer_size = sizeof sz_buffer;
		const auto result = RegQueryValueEx(
			key,
			TEXT("UpdateExeVolatile"),
			nullptr,
			nullptr,
			reinterpret_cast<LPBYTE>(sz_buffer),
			&dw_buffer_size);

		if (result == ERROR_SUCCESS && wcscmp(sz_buffer, L"0") != 0)
		{
			RegCloseKey(key);
			return true;
		}

		RegCloseKey(key);
	}

	const auto updatePendingKey = get_key(TEXT("SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\WindowsUpdate\\Services\\Pending"));
	if (updatePendingKey != nullptr)
	{
		DWORD total_subkeys = 0;

		RegQueryInfoKey(updatePendingKey,
			nullptr,
			nullptr,
			nullptr,
			&total_subkeys,
			nullptr,
			nullptr,
			nullptr,
			nullptr,
			nullptr,
			nullptr,
			nullptr);

		if (total_subkeys > 0)
		{
			RegCloseKey(updatePendingKey);
			return true;
		}

		RegCloseKey(updatePendingKey);
	}

	return false;
}

bool pending_reboot()
{
	return check_keys_exist() ||
		run_once_check() ||
		pending_update();
}

File: src/ProtonVPN.Core/OS/SystemState.cs
/*
 * Copyright (c) 2023 Proton AG
 *
 * This file is part of ProtonVPN.
 *
 * ProtonVPN is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ProtonVPN is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ProtonVPN.  If not, see <https://www.gnu.org/licenses/>.
 */

using System.Collections.Generic;
using System.Linq;
using Microsoft.Win32;

namespace ProtonVPN.Core.OS
{
    public class SystemState : ISystemState
    {
        private readonly List<string> _keysExist = new()
        {
            @"SOFTWARE\Microsoft\Windows\CurrentVersion\Component Based Servicing\RebootPending",
            @"SOFTWARE\Microsoft\Windows\CurrentVersion\Component Based Servicing\RebootInProgress",
            @"SOFTWARE\Microsoft\Windows\CurrentVersion\WindowsUpdate\Auto Update\RebootRequired",
            @"SOFTWARE\Microsoft\Windows\CurrentVersion\Component Based Servicing\PackagesPending",
            @"SOFTWARE\Microsoft\Windows\CurrentVersion\WindowsUpdate\Auto Update\PostRebootReporting",
        };

        private readonly Dictionary<string, List<string>> _valuesEqual = new()
        {
            {@"SOFTWARE\Microsoft\Windows\CurrentVersion\RunOnce", new List<string> {"DVDRebootSignal"}},
        };

        public bool PendingReboot()
        {
            if (_keysExist.Any(key => Registry.LocalMachine.OpenSubKey(key) != null))
            {
                return true;
            }

            if ((from item in _valuesEqual
                let key = Registry.LocalMachine.OpenSubKey(item.Key)
                let val = (string)key?.GetValue("")
                where val != null && item.Value.Contains(val)
                select item).Any())
            {
                return true;
            }

            return PendingUpdates();
        }

        private bool PendingUpdates()
        {
            RegistryKey key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Microsoft\Updates");
            object val = key?.GetValue("UpdateExeVolatile");
            if (val != null)
            {
                int.TryParse((string)val, out int result);
                if (result != 0)
                {
                    return true;
                }
            }

            key = Registry.LocalMachine.OpenSubKey(
                @"SOFTWARE\Microsoft\Windows\CurrentVersion\WindowsUpdate\Services\Pending");

            return key?.SubKeyCount > 0;
        }
    }
}